In [48]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf

# Read in the charity_data.csv
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")

application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [49]:
# Drop the non-beneficial ID columns, 'EIN', 'NAME', and ASK_AMT
application_df.drop(columns=['EIN'], inplace=True)
application_df

NAME APPLICATION_TYPE  \
0                           BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1                 AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2                     ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3                         SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4               GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   
...                                                  ...              ...   
34294              THE LIONS CLUB OF HONOLULU KAMEHAMEHA               T4   
34295           INTERNATIONAL ASSOCIATION OF LIONS CLUBS               T4   
34296                                PTA HAWAII CONGRESS               T3   
34297  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...               T5   
34298                           WATERHOUSE CHARITABLE TR               T3   

            AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0           Independent          C1000    ProductDev   Association       1   
1           Independent          C2000  Preservation  Co-operative       1   
2      CompanySponsored          C3000    ProductDev   Association       1   
3      CompanySponsored          C2000  Preservation         Trust       1   
4           Independent          C1000     Heathcare         Trust       1   
...                 ...            ...           ...           ...     ...   
34294       Independent          C1000    ProductDev   Association       1   
34295  CompanySponsored          C3000    ProductDev   Association       1   
34296  CompanySponsored          C2000  Preservation   Association       1   
34297       Independent          C3000    ProductDev   Association       1   
34298       Independent          C1000  Preservation  Co-operative       1   

          INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  IS_SUCCESSFUL  
0                  0                      N      5000              1  
1             1-9999                      N    108590              1  
2                  0                      N      5000              0  
3        10000-24999                      N      6692              1  
4      100000-499999                      N    142590              1  
...              ...                    ...       ...            ...  
34294              0                      N      5000              0  
34295              0                      N      5000              0  
34296              0                      N      5000              0  
34297              0                      N      5000              1  
34298          1M-5M                      N  36500179              0  

[34299 rows x 11 columns]

# Increase the application and classification type counts.

In [50]:
# Replace rare application types with "Other"
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
rare_application_types = application_type_counts[application_type_counts < 1000].index
application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(rare_application_types, "Other")


In [51]:
# Replace rare classifications with "Other"
classification_counts = application_df['CLASSIFICATION'].value_counts()
rare_classifications = classification_counts[classification_counts < 2000].index
application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(rare_classifications, "Other")


In [52]:
name_counts = application_df['NAME'].value_counts()
name_counts

NAME
PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: count, Length: 19568, dtype: int64

In [53]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
names_to_replace = list(name_counts[name_counts<5].index)
names_to_replace


# Replace in dataframe
for app in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['NAME'].value_counts()

NAME
Other                                                                19803
PARENT BOOSTER USA INC                                                1260
TOPS CLUB INC                                                          765
UNITED STATES BOWLING CONGRESS INC                                     700
WASHINGTON STATE UNIVERSITY                                            492
                                                                     ...  
GUILDS OF THE SANTA FE OPERA INC                                         5
FAMILY MOTOR COACH ASSOCIATION                                           5
JAMESTOWN SOCIETY                                                        5
MOST WORSHIPFUL PRINCE HALL GRAND LODGE OF FREE & ACCEPTED MASONS        5
PTA HAWAII CONGRESS                                                      5
Name: count, Length: 403, dtype: int64

In [66]:
# Convert categorical data to numeric with pd.get_dummies
application_df = pd.get_dummies(application_df, dtype=int)
application_df

STATUS   ASK_AMT  IS_SUCCESSFUL  NAME_AACE INTERNATIONAL  \
0           1      5000              1                        0   
1           1    108590              1                        0   
2           1      5000              0                        0   
3           1      6692              1                        0   
4           1    142590              1                        0   
...       ...       ...            ...                      ...   
34294       1      5000              0                        0   
34295       1      5000              0                        0   
34296       1      5000              0                        0   
34297       1      5000              1                        0   
34298       1  36500179              0                        0   

       NAME_ACE MENTOR PROGRAM OF AMERICA INC  NAME_ACTS MINISTRY  \
0                                           0                   0   
1                                           0                   0   
2                                           0                   0   
3                                           0                   0   
4                                           0                   0   
...                                       ...                 ...   
34294                                       0                   0   
34295                                       0                   0   
34296                                       0                   0   
34297                                       0                   0   
34298                                       0                   0   

       NAME_ACTS MISSIONS  \
0                       0   
1                       0   
2                       0   
3                       0   
4                       0   
...                   ...   
34294                   0   
34295                   0   
34296                   0   
34297                   0   
34298                   0   

       NAME_AFRICAN-AMERICAN POSTAL LEAGUE UNITED FOR SUCCESS A-PLUS  \
0                                                      0               
1                                                      0               
2                                                      0               
3                                                      0               
4                                                      0               
...                                                  ...               
34294                                                  0               
34295                                                  0               
34296                                                  0               
34297                                                  0               
34298                                                  0               

       NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  \
0                               0                                        0   
1                               0                                        0   
2                               0                                        0   
3                               0                                        0   
4                               0                                        0   
...                           ...                                      ...   
34294                           0                                        0   
34295                           0                                        0   
34296                           0                                        0   
34297                           0                                        0   
34298                           0                                        0   

       ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0      ...                  0                       0   
1      ...                  1                       0   
2      ...                  0                      

In [67]:
# Split preprocessed data into features and target arrays
X = application_df.drop(columns=['IS_SUCCESSFUL']).values
y = application_df['IS_SUCCESSFUL'].values

In [68]:
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [69]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Define and Compile the Model, 1st Optimization

In [70]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features =  X_train_scaled.shape[1]
hidden_nodes_layer1 =  33
hidden_nodes_layer2 = 28
hidden_nodes_layer3 = 18

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_21 (Dense)            (None, 33)                14586     
                                                                 
 dense_22 (Dense)            (None, 28)                952       
                                                                 
 dense_23 (Dense)            (None, 18)                522       
                                                                 
 dense_24 (Dense)            (None, 1)                 19        
                                                                 
Total params: 16079 (62.81 KB)
Trainable params: 16079 (62.81 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [62]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [63]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
858/858 [==============================] - 3s 2ms/step - loss: 0.4790 - accuracy: 0.7775
Epoch 2/100
858/858 [==============================] - 3s 3ms/step - loss: 0.4289 - accuracy: 0.7962
Epoch 3/100
858/858 [==============================] - 5s 6ms/step - loss: 0.4244 - accuracy: 0.7977
Epoch 4/100
858/858 [==============================] - 3s 4ms/step - loss: 0.4221 - accuracy: 0.7992
Epoch 5/100
858/858 [==============================] - 2s 3ms/step - loss: 0.4199 - accuracy: 0.8009
Epoch 6/100
858/858 [==============================] - 2s 2ms/step - loss: 0.4196 - accuracy: 0.7995
Epoch 7/100
858/858 [==============================] - 2s 2ms/step - loss: 0.4182 - accuracy: 0.8015
Epoch 8/100
858/858 [==============================] - 2s 2ms/step - loss: 0.4178 - accuracy: 0.8019
Epoch 9/100
858/858 [==============================] - 2s 2ms/step - loss: 0.4168 - accuracy: 0.8019
Epoch 10/100
858/858 [==============================] - 3s 3ms/step - loss: 0.4164 - accura

In [64]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - loss: 0.4729 - accuracy: 0.7857 - 448ms/epoch - 2ms/step
Loss: 0.47291067242622375, Accuracy: 0.7857142686843872


In [65]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharityOptimization.h5")

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


# Overview of the Analysis
The purpose of this analysis is to create a predictive model that can help Alphabet Soup identify applicatns for funding who are most likely to succeed. The aim is to build a binary classification model usign neural networks. The model should predict the likelihood of success for new applicants based on different feautures. This should ensure a better outcome for projects that need to be funded.

# Results

#### Data Preprocessing

1. What variable(s) are the target(s) for your model?

    * IS_SUCCESSFUL: This binary variable indicates whether the funding provided to an applicant was used effectively (1) or not (0).

2. What variable(s) are the features for your model?

    The features for our model are all the columns in the dataset except for the target variable and the identification columns. These features include:
    * APPLICATION_TYPE
    * AFFILIATION
    * CLASSIFICATION
    * USE_CASE
    * ORGANIZATION
    * STATUS
    * INCOME_AMT
    * SPECIAL_CONSIDERATIONS
    * ASK_AMT
    * NAME

3. What variable(s) should be removed from the input data because they are neither targets nor features?

    The variables that should be removed from the input data because they are neither targets nor features are:
    * EIN

# Summary

#### Overall Results of the Deep Learning Model

In this analysis, we aimed to build a binary classifier using a deep learning model to predict the success of applicants funded by Alphabet Soup. The model  achieved the accuracy of 78%, higher than the targeted accuracy. Here is a summary of the optimization step and the result:

1. First Optimization:
* Changes: Just dropped EIN column, set the test size to 20%, and included three hidden layers with 33, 28, and 18 neurons respectively.
* Accuracy: 78.57%
* Observations: Adding the NAMES column proved to be be a good choice compaired to the unoptimized version. This shows the importance of choosing the right feautures for the moddel.

#### Recommendation for a different model

Recommendation for a Different Model
Given that our deep learning model achieved the desired accuracy, it may still be beneficial to explore other models for comparison and validation. One potential alternative is the Random Forest Classifier.

##### Why Random Forest
1. It can handle categorical variables effectively without extensive preprocessing.
2. It averages the results of many decision trees which results in less risk of overfitting.
3. Provides understanding of which features are most predictive of success.
4. It is easier to use due to its ability of less tuning compared to deel learning models.